allennlp/models/encoder_decoders/nlvr/nlvr_semantic_parser.py

```diff
@@ -0,0 +1,614 @@
+        attention.
+    beam_size : ``int``
+    normalize_beam_score_by_length : ``bool``, optional (default=True)
+        Should the log probabilities be normalized by length before renormalizing them? Edunov et
+        al. do this in their work.
+    max_decoding_steps : ``int``
+        We use a beam search during decoding; what's the maximum number of steps we should take?
+    checklist_cost_weight : ``float``, optional (default=0.8)
+        Mixture weight (0-1) for combining coverage cost and denotation cost. As this increases, we
+        weigh the coverage cost higher, with a value of 1.0 meaning that we do not care about
+        denotation accuracy.
+    dynamic_cost_weight : ``Dict[str, Union[int, float]]``, optional (default=None)
+        A dict containing keys ``wait_num_epochs`` and ``rate`` indicating the number of steps
+        after which we should start decreasing the weight on checklist cost in favor of denotation
+        cost, and the rate at which we should do it. We will decrease the weight in the following
+        way - ``checklist_cost_weight = checklist_cost_weight - rate * checklist_cost_weight``
+        starting at the apropriate epoch.  The weight will remain constant if this is not provided.
+    penalize_non_agenda_actions : ``bool``, optional (default=False)
+        Should we penalize the model for producing terminal actions that are outside the agenda?
+    """
+    def __init__(self,
+                 vocab: Vocabulary,
+                 sentence_embedder: TextFieldEmbedder,
+                 nonterminal_embedder: TextFieldEmbedder,
+                 terminal_embedder: TextFieldEmbedder,
+                 encoder: Seq2SeqEncoder,
+                 attention_function: SimilarityFunction,
+                 beam_size: int,
+                 max_decoding_steps: int,
+                 normalize_beam_score_by_length: bool = True,
+                 checklist_cost_weight: float = 0.8,
+                 dynamic_cost_weight: Dict[str, Union[int, float]] = None,
+                 penalize_non_agenda_actions: bool = False) -> None:
+        super(NlvrSemanticParser, self).__init__(vocab=vocab)
+
+        self._sentence_embedder = sentence_embedder
+        self._denotation_accuracy = Average()
+        self._agenda_coverage = Average()
+        check_dimensions_match(nonterminal_embedder.get_output_dim(),
+                               terminal_embedder.get_output_dim(),
+                               "nonterminal embedding dim",
+                               "terminal embedding dim")
+        self._nonterminal_embedder = nonterminal_embedder
+        self._terminal_embedder = terminal_embedder
+        self._encoder = encoder
+        self._decoder_trainer: DecoderTrainer[Callable[[NlvrDecoderState], torch.Tensor]] = \
+                ExpectedRiskMinimization(beam_size, normalize_beam_score_by_length, max_decoding_steps)
```

Comment:
Why hard-code the trainer? Would it be difficult to make this general enough to pass the trainer to the init for training baseline models?

Reply:
Because it's not really feasible to swap out different trainers.  Taking these dependencies as inputs instead of instantiating them directly in the model is only useful when there are multiple valid alternatives and we want to experiment with them.  That's not true of our `DecoderTrainers`, and it's not clear to me that it will ever be true.  Also, because `DecoderTrainer` is generic over the supervision type, it only really makes sense to allow `DecoderTrainers` that match the supervision type that your model has, and we can't enforce that with our dependency injection.

If you find yourself wanting to try a few different training options, you'll first have to make sure that your model has the right supervision available for all of them.  Then you can add some flags to switch between them, but I'd still just instantiate them directly here.  _Maybe_ there will be a day when we have enough options and a very flexible model, and it will make sense to allow this dependency to be injected again.  But for now, it's just added complexity that isn't necessary.

Reply:
Agreed.